File: model_checking/BPjsModelChecking/src/main/resources/hot_cold.js
// if (typeof N !== 'undefined') { N = 3 }
// if (typeof M !== 'undefined') { M = 1 }
// if (typeof SOLVED !== 'undefined') { SOLVED = false }
// const N = 3
// const M = 1
// const SOLVED = true

const anyCold = bp.EventSet('anyCold', function (e) {
    return e.name.startsWith('COLD')
})

bp.registerBThread( "adda", function(){
    for (let i = 0; i < N; i++) {
        bp.hot(true).sync( {request:bp.Event("HOT")} );
    }
} );
for (let i = 0; i < M; i++) {
    bp.registerBThread("addb" + i.toString(), {i:i.toString()}, function(){
        for (let j = 0; j < N; j++) {
            bp.sync( {request:bp.Event("COLD" + bp.thread.data.i)} );
        }
    } );
}
if (SOLVED){
    bp.registerBThread( "control2", function(){
        let e = null;
        while (true){
            if (e != null && e.name.startsWith("COLD")){
                e = null;
                e = bp.sync( {waitFor:bp.all, block: anyCold} );
            } else {
                e = null;
                e = bp.sync( {waitFor:bp.all, block: bp.Event("HOT")} );
            }
        }
    } );
} else {
    bp.registerBThread( "control", function(){
        while (true){
            bp.sync( {waitFor:bp.Event("HOT")} );
            bp.sync( {waitFor:bp.all, block: bp.Event("HOT")} );
        }
    } );
}

File: model_checking/BPpyModelChecker/examples/hot_cold.py
#from bppy.model.b_thread import b_thread
from bppy.model.b_event import BEvent
#from bppy.model.sync_statement import waitFor, request, block
import bppy as bp

must_finish = "must_finish"
data = "data"


def set_bprogram(n, m):
    global N, M, event_list, any_portion, any_cold
    N = n
    M = m
    event_list = ["HOT"] + ["COLD" + str(i) for i in range(M)]
    any_portion = [BEvent("COLD" + str(i)) for i in range(M)] + [BEvent("HOT")]
    any_cold = [BEvent("COLD" + str(i)) for i in range(M)]


@bp.thread
def add_a():
    for i in range(N):
        yield bp.sync(request=BEvent("HOT"), must_finish=True, data=locals())


@bp.thread
def add_b(C):
    for i in range(N):
        yield bp.sync(request=BEvent(C), must_finish=False, data=locals())


@bp.thread
def control():
    while True:
        yield bp.sync(waitFor=BEvent("HOT"), must_finish=False, data=locals())
        yield bp.sync(waitFor=any_portion, block=BEvent("HOT"), must_finish=False, data=locals())


@bp.thread
def control2():
    e = yield bp.sync(waitFor=any_portion)
    while True:
        if e in any_cold:
            e = yield bp.sync(waitFor=any_portion, block=any_cold, must_finish=False)
        else:
            e = yield bp.sync(waitFor=any_portion, block=BEvent("HOT"), must_finish=False)




